lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/OperatorTopology.cs

```diff
@@ -161,14 +171,18 @@ public void OnNext(GroupCommunicationMessage gcm)
                 throw new ArgumentException("Message must have a source");
             }
 
-            NodeStruct sourceNode = FindNode(gcm.Source);
+            var sourceNode = FindNode(gcm.Source);
             if (sourceNode == null)
             {
                 throw new IllegalStateException("Received message from invalid task id: " + gcm.Source);
             }
 
-            _nodesWithData.Add(sourceNode);
-            sourceNode.AddData(gcm);
+            lock (_thisLock)
```

Comment:
Due to pipelining, now multiple messages from the same node are received and also sent to. As a result we cannot assume that from each node only single message is received. As a result, I have changed the logic where _nodesWithData is used only if the node is unable to find data from any other relevant node..otherwise it directly looks for data from the relevant node queue..... Hence adding node to _nodesWithData and adding it to relevant queue has to be done as one atomic step..as a result I put them as an atomic operation